docs: DOC-1141: Python version of `Configure JS plugins` (#8087)
deephaven/deephaven-core: docs/python/how-to-guides/configuration/js-plugins.md

---
title: Configure JS plugins
sidebar_label: JS plugins
---

The Deephaven server supports custom JS plugins that extend the functionality of the [server](https://github.com/deephaven/deephaven-core) and [web client UI](https://github.com/deephaven/web-client-ui). `server:latest` ships with several plugins pre-bundled, including Matplotlib, Plotly Express, and the Deephaven UI. This guide shows how to install a plugin that isn't included in the base image, using [Plotly](https://plotly.com/) (basic) as the example.

## Quickstart

Installing a JS plugin involves up to 3 steps:

```docker title="Dockerfile"
FROM ghcr.io/deephaven/web-plugin-packager:latest as js-plugins
# 1. Package the NPM deephaven-js-plugin(s)
RUN ./pack-plugins.sh <plugins>

FROM ghcr.io/deephaven/server:latest
# 2. Install the python js-plugin(s) if necessary (some plugins may be JS only)
RUN pip install --no-cache-dir <packages>
# 3. Copy the js-plugins/ directory
COPY --from=js-plugins js-plugins/ /opt/deephaven/config/js-plugins/
```

```bash
docker build -t my-deephaven-image .
docker run --rm -p 10000:10000 my-deephaven-image
```

> The workflow above is shown in relationship to the [Docker application](./docker-application.md), but the general requirements are the same if one is deploying the [production application](./configure-production-application.md). See [pack-plugins.sh](https://github.com/deephaven/deephaven-core/blob/main/docker/web-plugin-packager/src/main/docker/files/pack-plugins.sh) for more information on the JS plugin packaging logic.

Reviewer: ```suggestion
The workflow above applies to the [Docker application](./docker-application.md), but the general requirements are the same when deploying the [production application](./configure-production-application.md). See [`pack-plugins.sh`](https://github.com/deephaven/deephaven-core/blob/main/docker/web-plugin-packager/src/main/docker/files/pack-plugins.sh) for more information on the JS plugin packaging logic.
```

---
title: Configure JS plugins
sidebar_label: JS plugins
---

The Deephaven server supports custom JS plugins that extend the functionality of the [server](https://github.com/deephaven/deephaven-core) and [web client UI](https://github.com/deephaven/web-client-ui). This guide shows you how to install JS plugins and provides examples for two specific plugins, [Plotly](https://plotly.com/) and [Matplotlib](https://matplotlib.org/).

## Quickstart

Installing a JS plugin involves up to 3 steps:

```docker title="Dockerfile"
FROM ghcr.io/deephaven/web-plugin-packager:latest as js-plugins
# 1. Package the NPM deephaven-js-plugin(s)
RUN ./pack-plugins.sh <plugins>

FROM ghcr.io/deephaven/server:latest

> # 2. Install the python js-plugin(s) if necessary (some plugins may be JS only)

Reviewer: do we know if matplotlib / plotly is one of the groovy only ones?

RUN pip install --no-cache-dir <packages>
# 3. Copy the js-plugins/ directory
COPY --from=js-plugins js-plugins/ /opt/deephaven/config/js-plugins/
```

> <></>

Reviewer: stray brackets?

## Configuration

The JS plugins are automatically sourced from the `<configDir>/js-plugins/` directory if present. In the case of the Docker example above, `/opt/deephaven/config/` is the configuration directory. See [configuration directory](./configure-production-application.md#deephaven-server-bootstrap-configuration) for information about the configuration directory for other setups.

The JS plugins directory can also be set explicitly through the [configuration property](./config-file.md) `deephaven.jsPlugins.resourceBase`.

## Examples

### Plotly

Here's an example installing the [Plotly](https://plotly.com/) JS plugin:

```docker title="Dockerfile"
FROM ghcr.io/deephaven/web-plugin-packager:latest as js-plugins
RUN ./pack-plugins.sh @deephaven/js-plugin-plotly

FROM ghcr.io/deephaven/server:latest

> RUN pip install --no-cache-dir deephaven-plugin-plotly

Reviewer: can we test this to make sure it works properly in a Python console?
